### PR TITLE
03 authentication and authorization

### PR DESCRIPTION
**Pull request by aziarnik**

### CODE REVIEW

**Comment by aziarnik** [comment]
Password flow:
While user registers, password is created for him. I used bcrypt algorithm with salt and 11 rounds - it is long enough to make brute force attack harder, but user will not see difference in app working (one hashing password according to bcrypt library documentation should last 0,2 sec). I have also added pepper to it. I use it as private key of aes-256 algorithm to cipher hash. It is better approach then 'classic' use of pepper (adding it to password and hashing), because when pepper gets revealed, it is possible to generate new pepper, improve security of pepper and then decipher passwords with old pepper and cipher with new. In standard approach, this would be impossible (switch pepper to new one)

Authentication flow:
When user logs in, he gets the access token as jwt token with claims: id, username, role and refreshToken in authentication header. Refresh token (also jwtoken) is saved in separate db table called 'ActiveRefreshTokens'. When user' access token expires, he receives special error status, and can fire at auth/new-access-token endpoint to get new access token. There refresh token is checked, if any exist in db. If so, then:
1. User gets new accessToken with new refreshToken in it (this refresh token is saved to 'ActiveRefreshToken' table. His expiration date is the same as previous refresh token - it allows expiration of refresh token.
2. Used refresh token is archived (deleted from 'ActiveRefreshTokens' table and added to 'UsedRefreshTokens' table
3. When user tries to get new access token with old refresh token, then an event 'SomeoneUsedOldRefreshToken' is saved (then in job all active tokens for user are deleted), and special response status is sent.
4. If refreshtoken expired, then user will return 422 status, and has to log in again

This approach should decrease probability, that someone with stolen access token will gain access to user account.

When user logs out, all his refresh token used in accessToken is archived, and authentication headers are cleared.
When user tries to have access to guarded authorization endpoint, then only access token is validated (not refresh token). It is quicker solution then also checking refresh token (no additional db connection in invoking every endpoint), but less secure (active access token cannot be marked as fault)

Authorization:
There is separate middleware which checks access, based on user role and userid from access token (if not admin can edit/delete his own)

Tests:
I have created special docker compose file (docker.compose.test.yml) for tests (with separate volume). 
To launch test, first it is required to run compose:
npm run docker-compose-test-up
Then:

npm test

Tests works locally, here is my output:

Screenshot from 2022-10-31 12-55-14

File: src/db/repositories/base-repository-helper.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import { CustomError } from '../../errors/custom-error';
+import { errorMessages } from '../../errors/error-messages';
+import { ValueObject } from '../../value-objects/value-object-base';
+
+export class BaseRepositoryHelper {
+  static getRawColumnName(columnName: string): string {
+    return columnName.replace('"', '');
+  }
+
+  static objectToString(value: any): string {
```
Comment (Miszczu2137) [inline src/db/repositories/base-repository-helper.ts]: object isn't a perfect name for this function. I'd name it valueToString.

File: src/db/repositories/user-repository.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+  override async getAll(includeDeleted = false): Promise<User[]> {
+    if (includeDeleted) {
+      return super.getAll();
+    }
+    return (
+      await this.dbConnection.queryWithoutParams(
+        `SELECT ${this.columns}, ${this.systemColumns} FROM ${this.tableSchema}.${this.tableName} WHERE ${this.isNotDeletedContraint};`
+      )
+    ).map((x) => new User(x as IUserEntryData));
```
Comment (Miszczu2137) [inline src/db/repositories/user-repository.ts]: x => user

File: src/middlewares/authentication/authentication-settings.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import { RequestHandler } from 'express';
+import { Consts } from '../../consts';
+import { RequestHandlerHelper } from '../../helpers/request-handler-helper';
+import { AccessType } from './access-type';
+
+export const authenticationSettings = (opt: AccessType): RequestHandler => {
+  return (req, res, next) => {
+    const userContext = req.userContext;
+    const requestUserId = RequestHandlerHelper.getIdFromQueryParams(req);
+    switch (opt) {
```
Comment (Miszczu2137) [inline src/middlewares/authentication/authentication-settings.ts]: Please replace switch by if statements.

**Review by Miszczu2137 — APPROVED** [review]

**Pull request merged**